WNPRC_EHR/resources/queries/study/waterAmount.js

```diff
@@ -0,0 +1,125 @@
+
+    helper.registerRowProcessor(function (helper, row){
+        if (!row)
+            return;
+        if(!row.Id || !row.volume){
+            return;
+        }
+
+        var waterInForm = helper.getProperty('waterInForm');
+        waterInForm = waterInForm || {};
+        waterInForm[row.Id] = waterInForm[row.Id] || [];
+
+        var clientEncounterDate = helper.getProperty('clientEncounterDate');
+        clientEncounterDate = clientEncounterDate || {};
+        clientEncounterDate[row.Id] = clientEncounterDate [row.Id] || [];
+
+        if (row.objectid){
+            LABKEY.ExtAdapter.each(waterInForm[row.Id], function(r){
+                if (r.objectid == row.objectid){
+                    if(r.volume != row.volume){
+                        r.volume = row.volume;
+                    }
+                    else{
+                        return false;
+                    }
+                }
+            }, this)
+
+            LABKEY.ExtAdapter.each(clientEncounterDate[row.Id], function(r){
+                if (r.objectid == row.objectid){
+                    if(r.date != row.date){
+                        r.date = row.date;
+                    }
+                    else{
+                        return false;
+                    }
+                }
+            }, this)
+        }
+        helper.setProperty('waterInForm',waterInForm);
+        helper.setProperty('clientEncounterDate',clientEncounterDate);
+    });
+
+}
+
+function onUpsert(helper, scriptErrors, row, oldRow) {
+
+    if (row.date){
+        //console.log ("row date "+row.date );
+        var map = helper.getProperty('clientEncounterDate');
+        //console.log (map[row.Id]);
+        var clientEncounterDate = [];
+        if (map && map[row.Id]){
+            clientEncounterDate = map[row.Id];
+
+        }
+        let errorMessage = WNPRC.Utils.getJavaHelper().checkEncounterTime(row.Id,row.date,clientEncounterDate, 'waterAmount');
+        if (errorMessage != null){
+            EHR.Server.Utils.addError(scriptErrors,'date',errorMessage,'ERROR');
+        }
+
+    }
+
+    if (row.recordSource=="LabWaterForm"){
+        /*var today = new Date();
+        today.setHours(0,0,0,0);
+        var rowDate = new Date (row.date);
+        //rowDate.setHours(0,0,0,0);
+        if (rowDate.getTime() !== today.getTime()){
+            EHR.Server.Utils.addError (scriptErrors,'date','Additional water in this form can only be requested for the same date', 'ERROR');
+        }*/
+        //row.date = rowDate;
+        //console.log("value of qcstate waterAmount " + row.qcstate);
+        //console.log("value of date " + row.date);
+        var fixdate = new Date(row.date);
+        fixdate.setHours(14);
+        fixdate.setMinutes(1);
+        row.date = fixdate;
+        //console.log("value of fixdate " + row.date);
+        if (!row.skipWaterRegulationCheck){
+            //TODO: look up the qcstate given the label
+            row.qcstate = 10;
```

Comment:
@dnicolalde Would probably be best to look up the qc state given the label, something like EHR.Security.getQCStateMap() might do the trick.

Reply:
Change to what you are recommending.

Reply:
server side we have to use: EHR.Server.Security.getQCStateByLabel('Scheduled').Label or
EHR.Server.Security.getQCStateByLabel('Scheduled').rowid